File: tests/scalers/mongodb.test.ts
import * as async from 'async'
import * as fs from 'fs'
import * as sh from 'shelljs'
import * as tmp from 'tmp'
import test from 'ava'

const mongoDBNamespace = 'mongodb'
const testNamespace = 'mongodb-test'
const mongoDBUsername = 'test_user'
const mongoDBPassword = 'test_password'
const mongoDBDatabase = 'test'
const mongodbCollection = "test_collection"
const mongoJobName = "mongodb-job"

test.before(t => {
    // install mongoDB
    sh.exec(`kubectl create namespace ${mongoDBNamespace}`)
    const mongoDBTmpFile = tmp.fileSync()
    fs.writeFileSync(mongoDBTmpFile.name, mongoDBdeployYaml)

    t.is(0, sh.exec(`kubectl apply --namespace ${mongoDBNamespace} -f ${mongoDBTmpFile.name}`).code, 'creating a MongoDB deployment should work.')
    // wait for mongoDB to load
    let mongoDBReadyReplicaCount = '0'
    for (let i = 0; i < 30; i++) {
        mongoDBReadyReplicaCount = sh.exec(`kubectl get deploy/mongodb -n ${mongoDBNamespace} -o jsonpath='{.status.readyReplicas}'`).stdout
        if (mongoDBReadyReplicaCount != '1') {
            sh.exec('sleep 2s')
        }
    }
    t.is('1', mongoDBReadyReplicaCount, 'MongoDB is not in a ready state')

    const createUserJS = `db.createUser({ user:"${mongoDBUsername}",pwd:"${mongoDBPassword}",roles:[{ role:"readWrite", db: "${mongoDBDatabase}"}]})`
    const LoginJS = `db.auth("${mongoDBUsername}","${mongoDBPassword}")`

    const mongoDBPod = sh.exec(`kubectl get po -n ${mongoDBNamespace} -o jsonpath='{.items[0].metadata.name}'`).stdout
    t.not(mongoDBPod, '')
    sh.exec(`kubectl exec -n ${mongoDBNamespace} ${mongoDBPod} -- mongo --eval \'${createUserJS}\'`)
    sh.exec(`kubectl exec -n ${mongoDBNamespace} ${mongoDBPod} -- mongo --eval \'${LoginJS}\'`)

    sh.config.silent = true
    // create test namespace
    sh.exec(`kubectl create namespace ${testNamespace}`)

    // deploy streams consumer app, scaled job etc.
    const tmpFile = tmp.fileSync()
    const ConnectionString = Buffer.from(`mongodb://${mongoDBUsername}:${mongoDBPassword}@mongodb-svc.${mongoDBNamespace}.svc.cluster.local:27017/${mongoDBDatabase}`).toString()
    const base64ConnectionString = Buffer.from(`mongodb://${mongoDBUsername}:${mongoDBPassword}@mongodb-svc.${mongoDBNamespace}.svc.cluster.local:27017/${mongoDBDatabase}`).toString('base64')

    fs.writeFileSync(tmpFile.name, deployYaml.
    replace(/{{MONGODB_CONNECTION_STRING_BASE64}}/g, base64ConnectionString).
    replace(/{{MONGODB_JOB_NAME}}/g, mongoJobName).
    replace(/{{MONGODB_DATABASE}}/g, mongoDBDatabase).
    replace(/{{MONGODB_COLLECTION}}/g, mongodbCollection).
    replace(/{{MONGODB_CONNECTION_STRING}}/g,ConnectionString))

    t.is(
        0,
        sh.exec(`kubectl apply -f ${tmpFile.name} --namespace ${testNamespace}`).code,
        'creating a deployment should work..'
    )

})

test.serial('Job should have 0 job on start', t => {
    const jobCount = sh.exec(
        `kubectl get job --namespace ${testNamespace}`
    ).stdout
    t.is(jobCount, '', 'job count should start out as 0')
})

test.serial(`Job should scale to 5 then back to 0`, t => {
    // insert data to mongodb
    const InsertJS = `db.${mongodbCollection}.insert([
    {"region":"eu-1","state":"running","plan":"planA","goods":"apple"},
    {"region":"eu-1","state":"running","plan":"planA","goods":"orange"},
    {"region":"eu-1","state":"running","plan":"planA","goods":"strawberry"},
    {"region":"eu-1","state":"running","plan":"planA","goods":"cherry"},
    {"region":"eu-1","state":"running","plan":"planA","goods":"pineapple"}
    ])`
    const mongoDBPod = sh.exec(`kubectl get po -n ${mongoDBNamespace} -o jsonpath='{.items[0].metadata.name}'`).stdout
    t.not(mongoDBPod, '')

    t.is(
        0,
        sh.exec(`kubectl exec -n ${mongoDBNamespace} ${mongoDBPod} -- mongo --eval \'${InsertJS}\'`).code,
        'insert 5 mongo record'
    )

    let jobCount = '0'
    // maxJobCount = real Job + first line of output
    const maxJobCount = '6'

    for (let i = 0; i < 30 && jobCount !== maxJobCount; i++) {
        jobCount = sh.exec(
            `kubectl get job --namespace ${testNamespace} | wc -l`
        ).stdout.replace(/[\r\n]/g,"")

        if (jobCount !== maxJobCount) {
            sh.exec('sleep 2s')
        }
    }

    t.is(maxJobCount, jobCount, `Job count should be ${maxJobCount} after 60 seconds`)

    for (let i = 0; i < 36 && jobCount !== '0'; i++) {
        jobCount = sh.exec(
            `kubectl get job --namespace ${testNamespace} | wc -l`
        ).stdout.replace(/[\r\n]/g,"")
        if (jobCount !== '0') {
            sh.exec('sleep 5s')
        }
    }

    t.is('0', jobCount, 'Job count should be 0 after 3 minutes')
})

test.after.always.cb('clean up mongodb deployment', t => {
    const resources = [
        `scaledJob.keda.sh/${mongoJobName}`,
        'triggerauthentication.keda.sh/mongodb-trigger',
        `deployment.apps/mongodb`,
        'secret/mongodb-secret',
    ]

    for (const resource of resources) {
        sh.exec(`kubectl delete ${resource} --namespace ${testNamespace}`)
    }
    sh.exec(`kubectl delete namespace ${testNamespace}`)

    // uninstall mongodb
    sh.exec(`kubectl delete --namespace ${mongoDBNamespace} deploy/mongodb`)
    sh.exec(`kubectl delete namespace ${mongoDBNamespace}`)

    t.end()
})

const mongoDBdeployYaml = `
apiVersion: apps/v1
kind: Deployment
metadata:
  name: mongodb
spec:
  replicas: 1
  selector:
    matchLabels:
      name: mongodb
  template:
    metadata:
      labels:
        name: mongodb
    spec:
      containers:
      - name: mongodb
        image: mongo:4.2.1
        imagePullPolicy: IfNotPresent
        ports:
        - containerPort: 27017
          name: mongodb
          protocol: TCP
---
kind: Service
apiVersion: v1
metadata:
  name: mongodb-svc
spec:
  type: ClusterIP
  ports:
  - name: mongodb
    port: 27017
    targetPort: 27017
    protocol: TCP
  selector:
    name: mongodb
`

const deployYaml = `
apiVersion: keda.sh/v1alpha1
kind: TriggerAuthentication
metadata:
  name: mongodb-trigger
spec:
  secretTargetRef:
    - parameter: connectionString
      name: mongodb-secret
      key: connect
---
apiVersion: v1
kind: Secret
metadata:
  name: mongodb-secret
type: Opaque
data:
  connect: {{MONGODB_CONNECTION_STRING_BASE64}}
---
apiVersion: keda.sh/v1alpha1
kind: ScaledJob
metadata:
  name: {{MONGODB_JOB_NAME}}
spec:
  jobTargetRef:
    template:
      spec:
        containers:
          - name: mongodb-update
            image: 1314520999/mongodb-update:latest
            args:
            - --connectStr={{MONGODB_CONNECTION_STRING}}
            - --dataBase={{MONGODB_DATABASE}}
            - --collection={{MONGODB_COLLECTION}}
            imagePullPolicy: IfNotPresent
        restartPolicy: Never
    backoffLimit: 1
  pollingInterval: 20
  successfulJobsHistoryLimit: 0
  failedJobsHistoryLimit: 10
  triggers:
    - type: mongodb
      metadata:
        dbName: {{MONGODB_DATABASE}}
        collection: {{MONGODB_COLLECTION}}
        query: '{"region":"eu-1","state":"running","plan":"planA"}'
        queryValue: "1"
      authenticationRef:
        name: mongodb-trigger
---
`

File: tests/scalers/mongodb/mongodb_test.go
//go:build e2e
// +build e2e

package mongodb_test

import (
	"context"
	"encoding/base64"
	"fmt"
	"testing"
	"time"

	"github.com/joho/godotenv"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/client-go/kubernetes"

	. "github.com/kedacore/keda/v2/tests/helper"
)

// Load environment variables from .env file
var _ = godotenv.Load("../../.env")

const (
	testName = "mongodb-test"
)

var (
	testNamespace   = fmt.Sprintf("%s-ns", testName)
	secretName      = fmt.Sprintf("%s-secret", testName)
	triggerAuthName = fmt.Sprintf("%s-ta", testName)
	scaledJobName   = fmt.Sprintf("%s-sj", testName)
	mongoNamespace  = "mongo-ns"
	mongoUser       = "test_user"
	mongoPassword   = "test_password"
	mongoDBName     = "test"
	mongoCollection = "test_collection"
)

type templateData struct {
	TestNamespace                string
	MongoNamespace               string
	SecretName                   string
	TriggerAuthName              string
	ScaledJobName                string
	Connection, Base64Connection string
	Database, Collection         string
}

const (
	secretTemplate = `
apiVersion: v1
kind: Secret
metadata:
  name: {{.SecretName}}
  namespace: {{.TestNamespace}}
data:
  connectionString: {{.Base64Connection}}
`

	triggerAuthTemplate = `
apiVersion: keda.sh/v1alpha1
kind: TriggerAuthentication
metadata:
  name: {{.TriggerAuthName}}
  namespace: {{.TestNamespace}}
spec:
  secretTargetRef:
    - parameter: connectionString
      name: {{.SecretName}}
      key: connectionString
`

	scaledJobTemplate = `
apiVersion: keda.sh/v1alpha1
kind: ScaledJob
metadata:
  name: {{.ScaledJobName}}
  namespace: {{.TestNamespace}}
spec:
  jobTargetRef:
    template:
      spec:
        containers:
          - name: mongodb-update
            image: ghcr.io/kedacore/tests-mongodb:latest
            args:
            - --connectStr={{.Connection}}
            - --dataBase={{.Database}}
            - --collection={{.Collection}}
            imagePullPolicy: IfNotPresent
        restartPolicy: Never
    backoffLimit: 1
  pollingInterval: 20
  successfulJobsHistoryLimit: 0
  failedJobsHistoryLimit: 10
  triggers:
    - type: mongodb
      metadata:
        dbName: {{.Database}}
        collection: {{.Collection}}
        query: '{"region":"eu-1","state":"running","plan":"planA"}'
        queryValue: "1"
        activationQueryValue: "4"
      authenticationRef:
        name: {{.TriggerAuthName}}
`
)

func TestScaler(t *testing.T) {
	// setup
	t.Log("--- setting up ---")

	// Create kubernetes resources
	kc := GetKubernetesClient(t)
	data, templates := getTemplateData()
	t.Cleanup(func() {
		cleanupMongo(t)
		DeleteKubernetesResources(t, testNamespace, data, templates)
	})
	mongoPod := setupMongo(t, kc)
	CreateKubernetesResources(t, kc, testNamespace, data, templates)

	assert.True(t, WaitForScaledJobCount(t, kc, scaledJobName, testNamespace, 0, 60, 1),
		"job count should be 0 after 1 minute")

	// test scaling
	testActivation(t, kc, mongoPod)
	testScaleOut(t, kc, mongoPod)
}

func getTemplateData() (templateData, []Template) {
	connectionString := fmt.Sprintf("mongodb://%s:%s@mongodb.%s.svc.cluster.local:27017/%s",
		mongoUser, mongoPassword, mongoNamespace, mongoDBName)
	base64ConnectionString := base64.StdEncoding.EncodeToString([]byte(connectionString))

	return templateData{
			TestNamespace:    testNamespace,
			SecretName:       secretName,
			TriggerAuthName:  triggerAuthName,
			ScaledJobName:    scaledJobName,
			MongoNamespace:   mongoNamespace,
			Database:         mongoDBName,
			Collection:       mongoCollection,
			Connection:       connectionString,
			Base64Connection: base64ConnectionString,
		}, []Template{
			{Name: "secretTemplate", Config: secretTemplate},
			{Name: "triggerAuthTemplate", Config: triggerAuthTemplate},
			{Name: "scaledJobTemplate", Config: scaledJobTemplate},
		}
}

func setupMongo(t *testing.T, kc *kubernetes.Clientset) string {
	CreateNamespace(t, kc, mongoNamespace)
	_, err := ExecuteCommand(fmt.Sprintf("helm install mongodb --set auth.enabled=false --set persistence.enabled=false --namespace %s oci://registry-1.docker.io/cloudpirates/mongodb --wait", mongoNamespace))
	require.NoErrorf(t, err, "cannot execute command - %s", err)

	podList, err := kc.CoreV1().Pods(mongoNamespace).List(context.Background(), metav1.ListOptions{})
	require.NoErrorf(t, err, "cannot get mongo pod - %s", err)

	if len(podList.Items) != 1 {
		t.Error("cannot get mongo pod name")
		return ""
	}

	mongoPod := podList.Items[0].Name

	createUserCmd := fmt.Sprintf("db.createUser({ user:\"%s\",pwd:\"%s\",roles:[{ role:\"readWrite\", db: \"%s\"}]})",
		mongoUser, mongoPassword, mongoDBName)
	_, err = ExecuteCommand(fmt.Sprintf("kubectl exec %s -n %s -- mongosh --eval '%s'", mongoPod, mongoNamespace, createUserCmd))
	require.NoErrorf(t, err, "cannot create user - %s", err)

	loginCmd := fmt.Sprintf("db.auth(\"%s\",\"%s\")", mongoUser, mongoPassword)
	_, err = ExecuteCommand(fmt.Sprintf("kubectl exec %s -n %s -- mongosh --eval '%s'", mongoPod, mongoNamespace, loginCmd))
	require.NoErrorf(t, err, "cannot login - %s", err)

	return mongoPod
}

func testActivation(t *testing.T, kc *kubernetes.Clientset, mongoPod string) {
	t.Log("--- testing activation ---")

	insertCmd := fmt.Sprintf(`db.%s.insert([
		{"region":"eu-1","state":"running","plan":"planA","goods":"apple"},
		{"region":"eu-1","state":"running","plan":"planA","goods":"orange"}
		])`, mongoCollection)

	_, err := ExecuteCommand(fmt.Sprintf("kubectl exec %s -n %s -- mongosh --eval '%s'", mongoPod, mongoNamespace, insertCmd))
	assert.NoErrorf(t, err, "cannot insert mongo records - %s", err)
	assertNoScaledJobsDuring(t, kc, activationWindow)
}

// Two documents match the query and activationQueryValue is 4, so the ScaledJob must not create a
// job. The window is three of the ScaledJob's 20s polling intervals, so the scaler has evaluated the
// trigger against the inserted documents more than once before the test moves on. It is a window
// rather than a sleep followed by a count because successfulJobsHistoryLimit is 0: a job that was
// wrongly created and finished inside the window would have been removed by the time a single count
// ran, and would have consumed a document that the scale-out test expects to still be there.
const activationWindow = 60 * time.Second

func assertNoScaledJobsDuring(t *testing.T, kc *kubernetes.Clientset, window time.Duration) {
	t.Logf("Asserting that no job is created for %s for %s", scaledJobName, window)

	ctx, cancel := context.WithTimeout(context.Background(), window)
	defer cancel()

	err := KedaConsistently(ctx, func(ctx context.Context) (bool, error) {
		jobs, err := kc.BatchV1().Jobs(testNamespace).List(ctx, metav1.ListOptions{
			LabelSelector: fmt.Sprintf("scaledjob.keda.sh/name=%s", scaledJobName),
		})
		if err != nil {
			// KedaConsistently fails on a condition error, and a list that could not be made says
			// nothing about the jobs, so it is logged as an attempt that saw nothing to object to.
			t.Logf("cannot list jobs in namespace %s - %s", testNamespace, err)
			return true, nil
		}
		if len(jobs.Items) != 0 {
			return false, fmt.Errorf("%d job(s) exist for %s", len(jobs.Items), scaledJobName)
		}
		return true, nil
	}, IntervalShort)
	assert.NoErrorf(t, err, "job count should stay 0 for %s", window)
}

func testScaleOut(t *testing.T, kc *kubernetes.Clientset, mongoPod string) {
	t.Log("--- testing scale out ---")

	insertCmd := fmt.Sprintf(`db.%s.insert([
		{"region":"eu-1","state":"running","plan":"planA","goods":"strawberry"},
		{"region":"eu-1","state":"running","plan":"planA","goods":"cherry"},
		{"region":"eu-1","state":"running","plan":"planA","goods":"pineapple"}
		])`, mongoCollection)

	_, err := ExecuteCommand(fmt.Sprintf("kubectl exec %s -n %s -- mongosh --eval '%s'", mongoPod, mongoNamespace, insertCmd))
	assert.NoErrorf(t, err, "cannot insert mongo records - %s", err)

	assert.True(t, WaitForScaledJobCount(t, kc, scaledJobName, testNamespace, 5, 60, 1),
		"job count should be 5 after 1 minute")
}

func cleanupMongo(t *testing.T) {
	_, err := ExecuteCommand(fmt.Sprintf("helm uninstall mongodb --namespace %s", mongoNamespace))
	assert.NoErrorf(t, err, "cannot execute command - %s", err)
	DeleteNamespace(t, mongoNamespace)
}
